src/rgw/driver/sfs/sqlite/sqlite_multipart.cc

```diff
@@ -512,4 +512,71 @@ SQLiteMultipart::remove_multiparts_by_bucket_id_transact(
   return retry.run();
 }
 
+std::optional<DBDeletedMultipartItems>
+SQLiteMultipart::remove_done_or_aborted_multiparts_transact(uint max_items
+) const {
+  DBDeletedMultipartItems ret_parts;
+  auto storage = conn->get_storage();
+  RetrySQLite<DBDeletedMultipartItems> retry([&]() {
+    auto transaction = storage.transaction_guard();
+    // get first the list of parts to be deleted up to max_items
+    ret_parts = storage.select(
+        columns(
+            &DBMultipart::upload_id, &DBMultipart::object_uuid,
+            &DBMultipartPart::id
+        ),
+        inner_join<DBMultipart>(
+            on(is_equal(&DBMultipart::upload_id, &DBMultipartPart::upload_id))
+        ),
+        where(
+            is_equal(&DBMultipart::state, MultipartState::DONE) or
+            is_equal(&DBMultipart::state, MultipartState::ABORTED)
+        ),
+        order_by(&DBMultipartPart::id), limit(max_items)
+    );
+    auto ids = storage.select(
+        &DBMultipartPart::id,
+        inner_join<DBMultipart>(
+            on(is_equal(&DBMultipart::upload_id, &DBMultipartPart::upload_id))
+        ),
+        where(
+            is_equal(&DBMultipart::state, MultipartState::DONE) or
+            is_equal(&DBMultipart::state, MultipartState::ABORTED)
+        ),
+        order_by(&DBMultipartPart::id), limit(max_items)
+    );
+    ceph_assert(ids.size() == ret_parts.size());
```

Comment:
How about using ranges::transform to extract get<2>(row) into a std array instead of doing another query?

Reply:
I had to fix a couple of conflicts in the branch and added this suggestion.